File: Swift-C++-IR-Comparison/test/closures/Swift/EscapingAutoclosure.swift
// As an optimization, Swift may instead capture and store a copy of
//  a value if that value isn’t mutated by a closure, and if the value
//  isn’t mutated after the closure is created.

// Swift also handles all memory management involved in disposing of
// variables when they’re no longer needed.

@main
struct MyMain {

    static func main() {
        var customersInLine = [1, 2, 3, 4, 5, 6, 42]
        var customerProviders: [() -> Int] = []

        func collectProvider(_ customerProvider: @autoclosure @escaping () -> Int) {
            customerProviders.append(customerProvider)
        }

        collectProvider(customersInLine.remove(at: 0))
        collectProvider(customersInLine.remove(at: 0))

        for provider in customerProviders {
            let result = provider()
        }
    }
}

File: Swift-C++-Testsuite/AutoTranspiledTests/try-0/EscapingAutoclosure.cpp
 
#include<bits/stdc++.h>
using namespace std;

struct MyMain
{
    static void main()
    {
        vector<int> customersInLine = {1, 2, 3, 4, 5, 6, 42};
        vector<function<int()> > customerProviders;

		// Optimization - copy values if they are not mutated by a closure
        void collectProvider(const function<int()>& customerProvider)
        {
            customerProviders.push_back(customerProvider);
        }

        collectProvider([&](){return customersInLine.erase(customersInLine.begin());});
        collectProvider([&](){return customersInLine.erase(customersInLine.begin());});

        for(auto& provider : customerProviders)
        {
            int result = provider();
        }
    }
};

int main()
{
MyMain::main();
return 0;
}